equinox/nn/_batch_norm.py

```diff
@@ -10,7 +11,29 @@
 from ._stateful import State, StateIndex
 
 
-class BatchNorm(Module):
+class _StatefulLayer(Module):
```

Comment:
Let's put this in `nn/_stateful.py`, and additionally make it public API. (i.e. provide it in `__init__.py`). This should be something an end user could override in principle.

(I'm happy to hold off on documenting it for now, though.)